eZ/Publish/Core/Persistence/Cache/UrlWildcardHandler.php

```diff
@@ -0,0 +1,148 @@
+                'destinationUrl' => $destinationUrl,
+                'forward' => $forward,
+            ]
+        );
+
+        $urlWildcard = $this->persistenceHandler->urlWildcardHandler()->create($sourceUrl, $destinationUrl, $forward);
+
+        $this->cache->invalidateTags(['urlWildcard-notFound']);
+
+        return $urlWildcard;
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::remove
+     */
+    public function remove($id)
+    {
+        $this->logger->logCall(__METHOD__, ['id' => $id]);
+
+        $urlWildcard = $this->load($id);
+
+        $this->persistenceHandler->urlWildcardHandler()->remove($urlWildcard->id);
+
+        // need to clear lists of UrlWildcards cached due to loadAll()
+        $this->cache->invalidateTags(['ez-urlWildcard-id-' . $urlWildcard->id]);
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::load
+     */
+    public function load($id)
+    {
+        $this->logger->logCall(__METHOD__, ['id' => $id]);
+
+        $cacheItem = $this->cache->getItem('ez-urlWildcard-id-' . $id);
+
+        if ($cacheItem->isHit()) {
+            if (($return = $cacheItem->get()) === self::NOT_FOUND) {
+                throw new NotFoundException('UrlWildcard', $id);
+            }
+
+            return $return;
+        }
+
+        $urlWildcard = $this->persistenceHandler->urlWildcardHandler()->load($id);
+
+        $cacheItem->set($urlWildcard);
+        $cacheItem->tag($this->getCacheTags([$urlWildcard]));
+        $this->cache->save($cacheItem);
+
+        return $urlWildcard;
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::loadAll
+     */
+    public function loadAll($offset = 0, $limit = -1)
+    {
+        $this->logger->logCall(__METHOD__, ['offset' => $offset, 'limit' => $limit]);
+
+        return $this->persistenceHandler->urlWildcardHandler()->loadAll($offset, $limit);
+    }
+
+    /**
+     * @see \eZ\Publish\SPI\Persistence\Content\UrlWildcard\Handler::lookup
+     */
+    public function translate(string $sourceUrl): UrlWildcard
+    {
+        $this->logger->logCall(__METHOD__, ['url' => $sourceUrl]);
+
+        $cacheItem = $this->cache->getItem('ez-urlWildcard-source-' . $this->escapeForCacheKey($sourceUrl));
+
+        if ($cacheItem->isHit()) {
+            if (($return = $cacheItem->get()) === self::NOT_FOUND) {
+                throw new NotFoundException('UrlWildcard', $sourceUrl);
+            }
+
+            return $return;
+        }
+
+        $this->logger->logCacheMiss(['url' => $sourceUrl]);
+        try {
+            $urlWildcard = $this->persistenceHandler->urlWildcardHandler()->translate($sourceUrl);
+        } catch (APINotFoundException $e) {
+            $cacheItem->set(self::NOT_FOUND)
+                ->expiresAfter(30)
+                ->tag(['urlWildcard-notFound']);
+            $this->cache->save($cacheItem);
+            throw $e;
```

Comment:
I know we already do this, but I'm wondering if it might be better to wrap it in order to make sure line numbers and trace is kept correct on the original exception _(assuming the new throw tampers with that)_.

Reply:
Yes, that's a good approach. I improved it.